dev/sbt-plugin/src/main/scala/com/lightbend/lagom/sbt/LagomImport.scala

```diff
@@ -57,15 +57,14 @@ object LagomImport extends LagomImportCompat {
     fork in Test := true,
     concurrentRestrictions in Global += Tags.limit(Tags.Test, 1),
     javaOptions in Test ++= Seq("-Xms256M", "-Xmx512M"),
-    testGrouping in Test := singleTestsGrouping((definedTests in Test).value)
+    testGrouping in Test := singleTestsGrouping((definedTests in Test).value, (javaOptions in Test).value)
```

Comment:
Not sure if this is right actually. Isn't `javaOptions in (Test, test)` used for forking? (which will delegate to `javaOptions in Test` if undefined)

Reply:
I don't know. What about other test tasks like `testOnly`?

If it uses `javaOptions in (Test, test)`, then should is also be defining `javaOptions in (Test, test)` on the previous line? And `testGrouping in (Test, test)`, etc?

Reply:
…and `definedTests in (Test, test)`? I'm really not sure what's considered normal or best practice for sbt plugins.

Reply:
I think it depends on what's being defined by sbt, and what you are doing. You can take a look at `inspect test:executeTests` to find out what `test:test` transitively depends on. One of them is:
- `foo/test:test::javaOptions`

If you run `inspect test:test::javaOptions`, you'll see that it's provided by `*/*:javaOptions` (unscoped one).

As per where `javaOptions` should be defined, different options have pro and cons in terms of how opinionated your plugin becomes. We can think of different possibilities:

1. Use `globalSettings` and scope to nothing else (That's what the `Default.scala` does). This has the advantage of DRY if the user wants to override it, but it might not be what the user wants for non-Lagom projects.
2. Use `globalSettings` and scope to `Test`. Similar to above.
3. Use `projectSettings` and scope to `Test`. (What's done currently.) If the user wants to override it, they'd have to do it for all subprojects. Applies to `test`, `testOnly` etc.
4. Use `projectSettings` and scope to `(Test, test)`, this is most specific, which means if the user wants to override it it would need to be exactly that.

In the end, keeping 3 seems like a reasonable option. If the lookup is _specific_, and definition is _general_, this gives the build user to be more DRY. As in define one setting in their `build.sbt` and control multiple tasks.

Reply:
@eed3si9n thanks for the explanation.

It makes sense, but I think it still doesn't work the way you're saying, because of the default definition of `testGrouping` as derived from `javaOptions`, as we discussed previously.

When I make the change you suggest, if my project `hello-impl` defines:

```scala
javaOptions in Test += "-Djava.io.tmpdir=target/tmp"
```

I get the expected groupings:

```
> show hello-impl/test:test::testGrouping
[info] * Group(com.example.hello.impl.SomeOtherSpec,List(Test com.example.hello.impl.SomeOtherSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),None,Vector(-Xms256M, -Xmx512M, -Djava.io.tmpdir=target/tmp),false,Map())))
[info] * Group(com.example.hello.impl.HelloServiceSpec,List(Test com.example.hello.impl.HelloServiceSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),None,Vector(-Xms256M, -Xmx512M, -Djava.io.tmpdir=target/tmp),false,Map())))
[success] Total time: 1 s, completed 24/01/2018 4:57:48 PM
> show hello-impl/test:testOnly::testGrouping
[info] * Group(com.example.hello.impl.SomeOtherSpec,List(Test com.example.hello.impl.SomeOtherSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),None,Vector(-Xms256M, -Xmx512M, -Djava.io.tmpdir=target/tmp),false,Map())))
[info] * Group(com.example.hello.impl.HelloServiceSpec,List(Test com.example.hello.impl.HelloServiceSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),None,Vector(-Xms256M, -Xmx512M, -Djava.io.tmpdir=target/tmp),false,Map())))
[success] Total time: 0 s, completed 24/01/2018 4:57:53 PM
```

If it defines this instead:

```scala
javaOptions in (Test, test) += "-Djava.io.tmpdir=target/tmp"
```

...then the definition of `testGrouping` here is completely ignored in `test:test`:

```
> show hello-impl/test:test::testGrouping
[info] * Group(<default>,List(Test com.example.hello.impl.SomeOtherSpec : subclass(false, org.scalatest.Suite), Test com.example.hello.impl.HelloServiceSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),Some(/Users/tmoore/Projects/support/hpe/case-10181/hello/hello-impl),List(-Xms256M, -Xmx512M, -Djava.io.tmpdir=target/tmp),false,Map())))
```

Because it uses the default derived definition (which does include the updated `javaOptions`, but not the one-test-per-group configuration). This is surprising to me as a user (until you told me about how derived tasks work) but I'm not sure if there is anything we can really do about this other than changing sbt.

On the other hand, `testOnly` counter-intuitively uses the `javaOptions` definition that is supposed to be scoped to `(Test, test)`:

```
> show hello-impl/test:testOnly::testGrouping
[info] * Group(com.example.hello.impl.SomeOtherSpec,List(Test com.example.hello.impl.SomeOtherSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),None,Vector(-Xms256M, -Xmx512M, -Djava.io.tmpdir=target/tmp),false,Map())))
[info] * Group(com.example.hello.impl.HelloServiceSpec,List(Test com.example.hello.impl.HelloServiceSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),None,Vector(-Xms256M, -Xmx512M, -Djava.io.tmpdir=target/tmp),false,Map())))
```

Because that _is_ using Lagom's definition of `testGrouping`, which then looks for `javaOptions in (Test, test)` explicitly.

With the code as is in this commit, the same tests have the same results, except that `testOnly` does not include the modified `javaOptions` in its groups:

```
> show hello-impl/test:testOnly::testGrouping
[info] * Group(com.example.hello.impl.SomeOtherSpec,List(Test com.example.hello.impl.SomeOtherSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),None,Vector(-Xms256M, -Xmx512M),false,Map())))
[info] * Group(com.example.hello.impl.HelloServiceSpec,List(Test com.example.hello.impl.HelloServiceSpec : subclass(false, org.scalatest.Suite)),SubProcess(ForkOptions(None,None,List(),None,Vector(-Xms256M, -Xmx512M),false,Map())))
[success] Total time: 0 s, completed 24/01/2018 5:13:29 PM
```

This is more what I would expect: that if I defined `javaOptions` as scoped to `(Test, test)`, then it wouldn't apply to `testOnly`. Is my expectation right?

I'm not sure why you say that it's using `projectSettings` currently. This is just a method that exposes some settings for users to explicitly include into their projects if desired (opt in).

Reply:
I guess then you'd need to rewire `testGrouping` for all the scopes you care `(Test, test)` etc..

Reply:
I will merge it as it is now because is already an improvement over the previous situation. We'll tackle further improvements on another PR.